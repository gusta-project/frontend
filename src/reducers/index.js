import { combineReducers } from 'redux';

import {
  reducer as application,
  initialState as applicationState
} from './application';
import { reducer as profile, initialState as profileState } from './profile';
import { reducer as toast, initialState as toastState } from './toast';

export const initialState = {
  application: applicationState,
  profile: profileState,
  toast: toastState
};

export default combineReducers({
  application,
  profile,
  toast
});
