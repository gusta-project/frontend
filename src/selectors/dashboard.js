import { createSelector } from 'reselect';

export const getDashboard = (state) => state.dashboard;

export const getDashboardComponent = createSelector(
  getDashboard,
  (dashboard) => dashboard.dashboardComponent
);

export const getMigrations = createSelector(
  getDashboard,
  (dashboard) => dashboard.migrations
);

export const getStats = createSelector(
  getDashboard,
  (dashboard) => dashboard.stats
);
