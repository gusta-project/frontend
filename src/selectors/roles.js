import { createSelector } from 'reselect';

export const getRoles = (state) => state.roles;

export const getAllRoles = createSelector(
  getRoles,
  (roles) => roles.collection.roles
);

export const getCachedRoles = createSelector(
  getRoles,
  (roles) => roles.collection.cache
);

export const getRolesPager = createSelector(
  getRoles,
  (roles) => roles.collection.pager
);

export const getRoleUsers = createSelector(
  getRoles,
  (roles) => roles.roleUsers
);
