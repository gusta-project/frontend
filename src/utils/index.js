import dayjs from 'dayjs';

import { getApplication, getAuthorization } from 'selectors/application';

/* eslint-disable react/display-name, react/prop-types */

/**
 * Creates an object containing action constants namespaced under the specified reducer.
 *
 * @param {string} reducer The name of the reducer or module
 * @param {string[]} actions A list of action names
 * @return {object} Object with action name keys and full action string values
 */
export function buildActions(reducer, actions) {
  const result = {};

  for (const action of actions) {
    result[action] = `${reducer}/${action}`;
  }

  return result;
}

/**
 * Interpolate strings from the `params` object into the `url`.
 *
 * @param {object} endpoint Object with `url` and optional `params` object
 */
export const buildUrl = (endpoint) => {
  // this is defined by webpack.DefinePlugin
  const baseUrl = API_URL;
  const { url, params } = endpoint;

  if (!params) {
    return `${baseUrl}${url}`;
  }

  return Object.entries(params).reduce((resultUrl, entry) => {
    const [key, value] = entry;

    return resultUrl.replace(`{${key}}`, value);
  }, `${baseUrl}${url}`);
};

/**
 * Gets the initial Redux state from localStorage and the source code defaults.
 */
export const getInitialState = () => {
  // this needs to be require()d because an import results in a
  // circular dependency
  const { initialState } = require('reducers');

  try {
    const [accessToken, expiration] = [
      localStorage.getItem('accessToken'),
      localStorage.getItem('expiration')
    ];

    if (!accessToken) {
      return initialState;
    }

    const expirationDate = dayjs(JSON.parse(expiration));

    if (!expirationDate.isValid() || expirationDate.isBefore(dayjs())) {
      localStorage.removeItem('accessToken');
      localStorage.removeItem('expiration');
      return initialState;
    }

    return {
      ...initialState,
      application: {
        ...getApplication(initialState),
        authorization: {
          ...getAuthorization(initialState),
          accessToken: JSON.parse(accessToken),
          expiration: expirationDate
        }
      }
    };
  } catch {
    return initialState;
  }
};
