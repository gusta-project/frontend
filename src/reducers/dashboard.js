import { buildActions } from 'utils';

export const types = buildActions('dashboard', [
  'REQUEST_DASHBOARD',
  'REQUEST_DASHBOARD_SUCCESS',
  'SELECT_DASHBOARD',
  'SELECT_DASHBOARD_SUCCESS',
  'REQUEST_MIGRATIONS',
  'REQUEST_MIGRATIONS_SUCCESS',
  'REQUEST_MIGRATIONS_FAILURE',
  'REQUEST_STATS',
  'REQUEST_STATS_SUCCESS',
  'REQUEST_STATS_FAILURE'
]);

const requestDashboard = () => ({
  type: types.REQUEST_DASHBOARD
});

const requestDashboardSuccess = (dashboardComponent) => ({
  type: types.REQUEST_DASHBOARD_SUCCESS,
  dashboardComponent
});

const selectDashboard = ({ name, item }) => ({
  type: types.SELECT_DASHBOARD,
  name,
  item
});

const selectDashboardSuccess = (dashboardComponent) => ({
  type: types.SELECT_DASHBOARD_SUCCESS,
  dashboardComponent
});

const requestMigrations = () => ({
  type: types.REQUEST_MIGRATIONS
});

const requestMigrationsSuccess = (migrations) => ({
  type: types.REQUEST_MIGRATIONS_SUCCESS,
  migrations
});

const requestMigrationsFailure = (error) => ({
  type: types.REQUEST_MIGRATIONS_FAILURE,
  error
});

const requestStats = () => ({
  type: types.REQUEST_STATS
});

const requestStatsSuccess = (stats) => ({
  type: types.REQUEST_STATS_SUCCESS,
  stats
});

const requestStatsFailure = (error) => ({
  type: types.REQUEST_STATS_FAILURE,
  error
});

export const actions = {
  requestDashboard,
  requestDashboardSuccess,
  selectDashboard,
  selectDashboardSuccess,
  requestMigrations,
  requestMigrationsSuccess,
  requestMigrationsFailure,
  requestStats,
  requestStatsSuccess,
  requestStatsFailure
};

export const initialState = {
  dashboardComponent: {
    name: 'Home',
    item: null
  },
  migrations: [],
  stats: {}
};

export const reducer = (state = initialState, action = {}) => {
  switch (action.type) {
    case types.REQUEST_DASHBOARD_SUCCESS:
      return {
        ...state,
        dashboardComponent: action.dashboardComponent
      };
    case types.SELECT_DASHBOARD_SUCCESS:
      return {
        ...state,
        dashboardComponent: action.dashboardComponent
      };
    case types.REQUEST_MIGRATIONS_SUCCESS:
      return {
        ...state,
        migrations: action.migrations
      };
    case types.REQUEST_MIGRATIONS_FAILURE:
      return {
        ...state,
        error: action.error
      };
    case types.REQUEST_STATS_SUCCESS:
      return {
        ...state,
        stats: action.stats
      };
    case types.REQUEST_STATS_FAILURE:
      return {
        ...state,
        error: action.error
      };
    default:
      return state;
  }
};
