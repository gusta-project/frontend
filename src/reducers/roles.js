import { buildActions } from 'utils';

export const types = buildActions('roles', [
  'REQUEST_ROLES',
  'REQUEST_ROLES_SUCCESS',
  'CREATE_ROLE',
  'UPDATE_ROLE',
  'DELETE_ROLE',
  'REQUEST_ROLE_USERS',
  'REQUEST_ROLE_USERS_SUCCESS',
  'CREATE_ROLE_USER',
  'DELETE_ROLE_USER',
  'REQUEST_FAILURE',
  'CLEAR_COLLECTION'
]);

const requestRoles = (pager) => ({
  type: types.REQUEST_ROLES,
  pager
});

const requestRolesSuccess = (roles, pager) => ({
  type: types.REQUEST_ROLES_SUCCESS,
  roles,
  pager
});

const createRole = ({ name }) => ({
  type: types.CREATE_ROLE,
  name
});

const updateRole = ({ roleId, name }) => ({
  type: types.UPDATE_ROLE,
  roleId,
  name
});

const deleteRole = ({ roleId, name }) => ({
  type: types.DELETE_ROLE,
  roleId,
  name
});

const requestRoleUsers = ({ roleId }) => ({
  type: types.REQUEST_ROLE_USERS,
  roleId
});

const requestRoleUsersSuccess = (roleUsers) => ({
  type: types.REQUEST_ROLE_USERS_SUCCESS,
  roleUsers
});

const createRoleUser = ({ userId, roleId, active }) => ({
  type: types.CREATE_ROLE_USER,
  userId,
  roleId,
  active
});

const deleteRoleUser = ({ userId, roleId, name }) => ({
  type: types.DELETE_ROLE_USER,
  userId,
  roleId,
  name
});

const requestFailure = (error) => ({
  type: types.REQUEST_FAILURE,
  error
});

const clearCollection = () => ({
  type: types.CLEAR_COLLECTION
});

export const actions = {
  requestRoles,
  requestRolesSuccess,
  createRole,
  updateRole,
  deleteRole,
  requestRoleUsers,
  requestRoleUsersSuccess,
  createRoleUser,
  deleteRoleUser,
  requestFailure,
  clearCollection
};

export const initialState = {
  error: null,
  collection: {
    cache: [],
    roles: [],
    pager: {
      count: null,
      limit: 100,
      page: 1,
      pages: null
    }
  },
  roleUsers: []
};

export const reducer = (state = initialState, action = {}) => {
  switch (action.type) {
    case types.REQUEST_ROLES_SUCCESS:
      return {
        ...state,
        collection: {
          cache: action.roles,
          roles: action.roles[action.pager.page],
          pager: {
            ...state.pager,
            ...action.pager
          }
        }
      };
    case types.REQUEST_ROLE_USERS_SUCCESS:
      return {
        ...state,
        roleUsers: action.roleUsers
      };
    case types.REQUEST_FAILURE:
      return {
        ...state,
        error: action.error
      };
    case types.CLEAR_COLLECTION:
      return {
        ...state,
        collection: {
          ...initialState.collection
        }
      };
    default:
      return state;
  }
};
