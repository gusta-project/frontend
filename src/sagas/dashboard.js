import { all, call, put, takeLatest } from 'redux-saga/effects';

import request from 'utils/request';
import { actions, types } from 'reducers/dashboard';
import { actions as toastActions } from 'reducers/toast';

function* requestDashboardWorker(dashboardComponent) {
  yield put(actions.requestDashboardSuccess(dashboardComponent));
}

function* selectDashboardWorker({ name, item }) {
  yield put(actions.selectDashboardSuccess({ name, item }));
}

function* requestMigrationsWorker() {
  try {
    const endpoint = {
      url: '/data/version',
      method: 'GET'
    };
    const result = yield call(request.execute, { endpoint });

    // update migrations in state or throw an error
    if (result.success) {
      const {
        response: { data }
      } = result;

      yield put(actions.requestMigrationsSuccess(data));
    } else if (result.error) {
      throw result.error;
    } else {
      throw new Error('Failed to get migrations!');
    }
  } catch (error) {
    const { message } = error;

    yield put(actions.requestMigrationsFailure(error));
    yield put(
      toastActions.popToast({
        title: 'Error',
        icon: 'times-circle',
        message
      })
    );
  }
}

function* requestStatsWorker() {
  try {
    const endpoint = {
      url: '/stats/dashboard',
      method: 'GET'
    };
    const result = yield call(request.execute, { endpoint });

    // update stats in state or throw an error
    if (result.success) {
      const {
        response: { data }
      } = result;

      yield put(actions.requestStatsSuccess(data));
    } else if (result.error) {
      throw result.error;
    } else {
      throw new Error('Failed to get statistics!');
    }
  } catch (error) {
    const { message } = error;

    yield put(actions.requestStatsFailure(error));
    yield put(
      toastActions.popToast({
        title: 'Error',
        icon: 'times-circle',
        message
      })
    );
  }
}

function* requestDashboardWatcher() {
  yield takeLatest(types.REQUEST_DASHBOARD, requestDashboardWorker);
}

function* selectDashboardWatcher() {
  yield takeLatest(types.SELECT_DASHBOARD, selectDashboardWorker);
}

function* requestMigrationsWatcher() {
  yield takeLatest(types.REQUEST_MIGRATIONS, requestMigrationsWorker);
}

function* requestStatsWatcher() {
  yield takeLatest(types.REQUEST_STATS, requestStatsWorker);
}

export const workers = {
  requestDashboardWorker,
  selectDashboardWorker,
  requestMigrationsWorker,
  requestStatsWorker
};

export const watchers = {
  requestDashboardWatcher,
  selectDashboardWatcher,
  requestMigrationsWatcher,
  requestStatsWatcher
};

export default function* saga() {
  yield all(Object.values(watchers).map(watcher => watcher()));
}
