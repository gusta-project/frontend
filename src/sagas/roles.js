import { all, call, put, takeLatest, select } from 'redux-saga/effects';
import request from 'utils/request';
import helper from 'utils/saga';
import { getCachedRoles, getRolesPager } from 'selectors/roles';
import { actions, types } from 'reducers/roles';

function* requestRolesWorker({ pager }) {
  try {
    const cached = yield select(getCachedRoles);
    const store = yield select(getRolesPager);

    const response = yield call(helper.pager, {
      cached,
      pager: {
        ...pager,
        store
      },
      route: {
        count: '/roles/count',
        data: '/roles/'
      },
      type: 'Roles'
    });

    yield put(actions.requestRolesSuccess(response.cached, response.pager));
  } catch (error) {
    const { message } = error;

    yield put(actions.requestFailure(error));
    yield call(helper.errorToast, message);
  }
}

function* createRoleWorker({ name }) {
  try {
    const endpoint = {
      url: '/role',
      method: 'POST'
    };
    const data = {
      name
    };
    const result = yield call(request.execute, { endpoint, data });

    // update roles in state or throw an error
    if (result.success) {
      yield put(actions.clearCollection());
      yield put(actions.requestRoles({ page: 1, limit: 20 }));
      yield call(helper.toast, {
        title: 'Role Created',
        message: `${name} role successfully created!`
      });
    } else if (result.error) {
      throw result.error;
    } else {
      throw new Error('Failed to create role!');
    }
  } catch (error) {
    const { message } = error;

    yield put(actions.requestFailure(error));
    yield call(helper.errorToast, message);
  }
}

function* updateRoleWorker({ roleId, name }) {
  try {
    const endpoint = {
      url: `/role/${roleId}`,
      method: 'PUT'
    };
    const data = {
      name
    };
    const result = yield call(request.execute, { endpoint, data });

    // update roles in state or throw an error
    if (result.success) {
      yield put(actions.clearCollection());
      yield put(actions.requestRoles({ page: 1, limit: 20 }));
      yield call(helper.toast, {
        title: 'Edit Role',
        message: `${name} role successfully updated!`
      });
    } else if (result.error) {
      throw result.error;
    } else {
      throw new Error('Failed to update role!');
    }
  } catch (error) {
    const { message } = error;

    yield put(actions.requestFailure(error));
    yield call(helper.errorToast, message);
  }
}

function* deleteRoleWorker({ roleId, name }) {
  try {
    const endpoint = {
      url: `/role/${roleId}`,
      method: 'DELETE'
    };
    const result = yield call(request.execute, { endpoint });

    // update roles in state or throw an error
    if (result.success) {
      yield put(actions.clearCollection());
      yield put(actions.requestRoles({ page: 1, limit: 20 }));
      yield call(helper.toast, {
        title: 'Delete Role',
        message: `${name} role successfully deleted!`
      });
    } else if (result.error) {
      throw result.error;
    } else {
      throw new Error('Failed to delete role!');
    }
  } catch (error) {
    const { message } = error;

    yield put(actions.requestFailure(error));
    yield call(helper.errorToast, message);
  }
}

function* requestRoleUsersWorker({ roleId }) {
  try {
    const endpoint = {
      url: `/users/role/${roleId}`,
      method: 'GET'
    };
    const result = yield call(request.execute, { endpoint });

    // update role users in state or throw an error
    if (result.success) {
      const {
        response: { data }
      } = result;

      yield put(actions.requestRoleUsersSuccess(data));
    } else if (result.error) {
      throw result.error;
    } else {
      throw new Error('Failed to get role users!');
    }
  } catch (error) {
    const { message } = error;

    yield put(actions.requestFailure(error));
    yield call(helper.errorToast, message);
  }
}

function* createRoleUserWorker({ userId, roleId, active }) {
  try {
    const endpoint = {
      url: `/user/${userId}/role`,
      method: 'POST'
    };
    const data = {
      userId,
      roleId,
      active
    };
    const result = yield call(request.execute, { endpoint, data });

    // update roles in state or throw an error
    if (result.success) {
      yield call(helper.toast, {
        title: 'User Role Added',
        message: `Role ${roleId} assigned to User ${userId}!`
      });
    } else if (result.error) {
      throw result.error;
    } else {
      throw new Error('Failed to assign role to user!');
    }
  } catch (error) {
    const { message } = error;

    yield put(actions.requestFailure(error));
    yield call(helper.errorToast, message);
  }
}

function* deleteRoleUserWorker({ userId, roleId, name }) {
  try {
    const endpoint = {
      url: `/user/${userId}/role/${roleId}`,
      method: 'DELETE'
    };
    const result = yield call(request.execute, { endpoint });

    // update roles in state or throw an error
    if (result.success) {
      yield call(helper.toast, {
        title: 'Unassign User Role',
        message: `${name} role successfully unassigned!`
      });
    } else if (result.error) {
      throw result.error;
    } else {
      throw new Error('Failed to unassign role!');
    }
  } catch (error) {
    const { message } = error;

    yield put(actions.requestFailure(error));
    yield call(helper.errorToast, message);
  }
}

function* requestRolesWatcher() {
  yield takeLatest(types.REQUEST_ROLES, requestRolesWorker);
}

function* createRoleWatcher() {
  yield takeLatest(types.CREATE_ROLE, createRoleWorker);
}

function* updateRoleWatcher() {
  yield takeLatest(types.UPDATE_ROLE, updateRoleWorker);
}

function* deleteRoleWatcher() {
  yield takeLatest(types.DELETE_ROLE, deleteRoleWorker);
}

function* requestRoleUsersWatcher() {
  yield takeLatest(types.REQUEST_ROLE_USERS, requestRoleUsersWorker);
}

function* createRoleUserWatcher() {
  yield takeLatest(types.CREATE_ROLE_USER, createRoleUserWorker);
}

function* deleteRoleUserWatcher() {
  yield takeLatest(types.DELETE_ROLE_USER, deleteRoleUserWorker);
}

export const workers = {
  requestRolesWorker,
  createRoleWorker,
  updateRoleWorker,
  deleteRoleWorker,
  requestRoleUsersWorker,
  createRoleUserWorker,
  deleteRoleUserWorker
};

export const watchers = {
  requestRolesWatcher,
  createRoleWatcher,
  updateRoleWatcher,
  deleteRoleWatcher,
  requestRoleUsersWatcher,
  createRoleUserWatcher,
  deleteRoleUserWatcher
};

export default function* saga() {
  yield all(Object.values(watchers).map(watcher => watcher()));
}
