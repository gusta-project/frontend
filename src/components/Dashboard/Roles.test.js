import React from 'react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';

import { initialState } from 'reducers/roles';
import { initialState as dashboardInitialState } from 'reducers/dashboard';
import ConnectedRoles from './Roles';
import { withMemoryRouter } from 'utils';

describe('Dashboard <Roles />', () => {
  const defaultLayoutOptions = {
    border: false,
    header: true,
    title: false,
    style: {}
  };
  const mockStore = configureStore();
  const store = mockStore({
    roles: initialState,
    dashboard: dashboardInitialState
  });

  const RoutedRoles = withMemoryRouter(ConnectedRoles);

  it('renders correctly', () => {
    expect(
      renderer
        .create(
          <Provider store={store}>
            <RoutedRoles layoutOptions={defaultLayoutOptions} />
          </Provider>
        )
        .toJSON()
    ).toMatchSnapshot();
  });
});
