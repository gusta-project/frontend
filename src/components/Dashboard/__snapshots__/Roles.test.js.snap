// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard <Roles /> renders correctly 1`] = `<Pagination />`;
